cpp/tests/rolling/rolling_test.cu

```diff
@@ -0,0 +1,502 @@
+    cudf::test::column_wrapper<T> ref(ref_data, [&](gdf_index_type i) { return ref_data_valid[i]; } );
+
+    // print the columns for debugging
+    //out.print();
+    //ref.print();
+
+    ASSERT_TRUE(out == ref);
+  }
+
+  void run_test(gdf_size_type w, gdf_size_type m, gdf_size_type f,
+		std::vector<gdf_size_type> &window, std::vector<gdf_size_type> &min_periods, std::vector<gdf_size_type> &forward_window,
+		gdf_agg_op agg)
+  {
+    create_gdf_input_buffers();
+
+    gdf_size_type *d_window = NULL;
+    gdf_size_type *d_min_periods = NULL;
+    gdf_size_type *d_forward_window = NULL;
+
+    // copy sizes to the gpu
+    if (window.size() > 0) {
+      EXPECT_EQ(RMM_ALLOC(&d_window, window.size() * sizeof(gdf_size_type), 0), RMM_SUCCESS);
+      CUDA_TRY(cudaMemcpy(d_window, window.data(), window.size() * sizeof(gdf_size_type), cudaMemcpyDefault));
+    }
+    if (min_periods.size() > 0) {
+      EXPECT_EQ(RMM_ALLOC(&d_min_periods, min_periods.size() * sizeof(gdf_size_type), 0), RMM_SUCCESS);
+      CUDA_TRY(cudaMemcpy(d_min_periods, min_periods.data(), min_periods.size() * sizeof(gdf_size_type), cudaMemcpyDefault));
+    }
+    if (forward_window.size() > 0) {
+      EXPECT_EQ(RMM_ALLOC(&d_forward_window, forward_window.size() * sizeof(gdf_size_type), 0), RMM_SUCCESS);
+      CUDA_TRY(cudaMemcpy(d_forward_window, forward_window.data(), forward_window.size() * sizeof(gdf_size_type), cudaMemcpyDefault));
+    }
+
+    out_gdf_col = { cudf::rolling_window(in_gdf_col.get(), w, m, f, agg, d_window, d_min_periods, d_forward_window), deleter };
+
+    create_reference_output(agg, w, m, f, window, min_periods, forward_window);
+
+    compare_gdf_result();
+  
+    // free GPU memory 
+    if (d_window != NULL) EXPECT_EQ(RMM_FREE(d_window, 0), RMM_SUCCESS);
+    if (d_min_periods != NULL) EXPECT_EQ(RMM_FREE(d_min_periods, 0), RMM_SUCCESS);
+    if (d_forward_window != NULL) EXPECT_EQ(RMM_FREE(d_forward_window, 0), RMM_SUCCESS);
+  }
+
+  template<class... TArgs>
+  void run_test_all_agg(TArgs... FArgs)
+  {
+    // test all supported aggregators
+    run_test(FArgs..., GDF_SUM);
+    run_test(FArgs..., GDF_MIN);
+    run_test(FArgs..., GDF_MAX);
+    run_test(FArgs..., GDF_COUNT);
+    run_test(FArgs..., GDF_AVG);
+    
+    // this aggregation function is not supported yet - expected to throw an exception
+    EXPECT_THROW(run_test(FArgs..., GDF_COUNT_DISTINCT), cudf::logic_error);
+  }
+
+  // input
+  std::vector<T> in_col;
+  std::vector<bool> in_col_valid;
+  gdf_col_pointer in_gdf_col;
+
+  // reference
+  std::vector<T> ref_data;
+  std::vector<bool> ref_data_valid;
+
+  // output
+  gdf_col_pointer out_gdf_col;
+
+  // column deleter
+  const std::function<void(gdf_column*)> deleter = [](gdf_column* col) {
+    col->size = 0;
+    RMM_FREE(col->data, 0);
+    RMM_FREE(col->valid, 0);
+  };
+};
+
+using TestTypes = ::testing::Types<int8_t, int16_t, int32_t, int64_t, double>;
+
+TYPED_TEST_CASE(RollingTest, TestTypes);
```

Comment:
Please also add a test to exercise all possible failure cases (any `CUDF_EXPECTS` that could fail (perhaps excluding allocation failures), should have a test that makes it fail and asserts if it doesn't.  

This test doesn't need to be a TYPED_TEST.

Reply:
I have added a test for negative sizes, and I have a test that checks the unsupported aggregator (COUNT_DISTINCT). There are a few other combinations we could check that suppose to fail and I've added a list to the end of the file, not sure if I have time to implement all that, but we can add those tests later.

Reply:
Since there are no python bindings yet, I'd rather slip the release than have loose ends.